# <span style="color:red"> Lecture 18a: Pandas Operations and Data Manipulation </span>

<font size = "4">

We finish up the content from Lecture 17 we didn't get to last time.

## <span style="color:red"> I. Import Libraries and Read in Files</span>

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [3]:
# Read in the car features dataset

carfeatures = pd.read_csv("data/features.csv")

## <span style="color:red"> II. Subsetting by row/column position </span>

<font size = "4">

- In assignment #5, we saw how to use the `.loc` attribute to grab the value of a variable from a given row. We'll come back to that method in a later lecture.

- In addition, each DataFrame has an "iloc" attribute. This stands for "integer-location based indexing".

- **Important Note:** `.iloc` is an attribute, not a method! You use square brackets `[]` not parentheses!

In [7]:
print(carfeatures)
print(carfeatures.columns.values)

      mpg  cylinders  displacement horsepower  weight  acceleration vehicle_id
0    18.0          8           307        130    3504          12.0  C-1689780
1    15.0          8           350        165    3693          11.5  B-1689791
2    18.0          8           318        150    3436          11.0  P-1689802
3    16.0          8           304        150    3433          12.0  A-1689813
4    17.0          8           302        140    3449          10.5  F-1689824
..    ...        ...           ...        ...     ...           ...        ...
393  27.0          4           140         86    2790          15.6  F-1694103
394  44.0          4            97         52    2130          24.6  V-1694114
395  32.0          4           135         84    2295          11.6  D-1694125
396  28.0          4           120         79    2625          18.6  F-1694136
397  31.0          4           119         82    2720          19.4  C-1694147

[398 rows x 7 columns]
['mpg' 'cylinders' 'displace

<font size = "4">

- Above, see that car 0 has 18.0 mpg

- The value "18.0" is in the first row, and corresponds to the first column (the index column on the left is NOT considered a column/variable)

- Since Python starts counting from zero, this is in row 0 column 0

- In the next cell we access this value using the `.iloc` attribute.


In [8]:
# value in row 0 col 0
car0_mpg = carfeatures.iloc[0,0]
print(car0_mpg)

18.0


<font size = "4">

- Looking at `carfeatures`, we see that car 3 has a weight of 3433

- Counting from zero, the value "3433" is in row 3, column 4

In [9]:
car3_weight = carfeatures.iloc[3, 4]
print(car3_weight)

3433


<font size = "4">

- Let's grab **all** the features for car 0 (mpg, cylinders, displacement, ..., vehicle_id)

- So we want all the entries belonging to row zero.

- Use a ":" (colon) for the column index to grab **all** the columns

In [11]:
car0_features = carfeatures.iloc[0,:]
print(car0_features)
print()
print(type(car0_features))

mpg                  18.0
cylinders               8
displacement          307
horsepower            130
weight               3504
acceleration         12.0
vehicle_id      C-1689780
Name: 0, dtype: object

<class 'pandas.core.series.Series'>


<font size = "4">

- Grab all columns for car 3

In [12]:
car3_features = carfeatures.iloc[3,:]
print(car3_features)

mpg                  16.0
cylinders               8
displacement          304
horsepower            150
weight               3433
acceleration         12.0
vehicle_id      A-1689813
Name: 3, dtype: object


In [13]:
print(carfeatures)

      mpg  cylinders  displacement horsepower  weight  acceleration vehicle_id
0    18.0          8           307        130    3504          12.0  C-1689780
1    15.0          8           350        165    3693          11.5  B-1689791
2    18.0          8           318        150    3436          11.0  P-1689802
3    16.0          8           304        150    3433          12.0  A-1689813
4    17.0          8           302        140    3449          10.5  F-1689824
..    ...        ...           ...        ...     ...           ...        ...
393  27.0          4           140         86    2790          15.6  F-1694103
394  44.0          4            97         52    2130          24.6  V-1694114
395  32.0          4           135         84    2295          11.6  D-1694125
396  28.0          4           120         79    2625          18.6  F-1694136
397  31.0          4           119         82    2720          19.4  C-1694147

[398 rows x 7 columns]


<font size = "4">

- Column 0 corresponds to "mpg" (the index column on the far left doesn't count)

- Grab the entire column 0, which means grab **all** the rows in this column

In [14]:
car_mpg = carfeatures.iloc[:, 0]
print(car_mpg)
print()
print(type(car_mpg))

0      18.0
1      15.0
2      18.0
3      16.0
4      17.0
       ... 
393    27.0
394    44.0
395    32.0
396    28.0
397    31.0
Name: mpg, Length: 398, dtype: float64

<class 'pandas.core.series.Series'>


<font size = "4">

- Horsepower is column 3

- Grab **all** rows of column 3

In [15]:
car_horsepower = carfeatures.iloc[:, 3]
display(car_horsepower)

0      130
1      165
2      150
3      150
4      140
      ... 
393     86
394     52
395     84
396     79
397     82
Name: horsepower, Length: 398, dtype: object

<font size = "4">

- Grab all features for cars 0, 4, 396 (so rows 0, 4, 396)

In [16]:
cargroup_features = carfeatures.iloc[[0, 4, 396],:]
print(cargroup_features)
print(type(cargroup_features))

      mpg  cylinders  displacement horsepower  weight  acceleration vehicle_id
0    18.0          8           307        130    3504          12.0  C-1689780
4    17.0          8           302        140    3449          10.5  F-1689824
396  28.0          4           120         79    2625          18.6  F-1694136
<class 'pandas.core.frame.DataFrame'>


In [17]:
# Grab all features for cars 0, 4, 396 (so rows 0, 4, 396)

group_rows = [0, 4, 396]
cargroup_features = carfeatures.iloc[ group_rows, : ]
print(cargroup_features)

      mpg  cylinders  displacement horsepower  weight  acceleration vehicle_id
0    18.0          8           307        130    3504          12.0  C-1689780
4    17.0          8           302        140    3449          10.5  F-1689824
396  28.0          4           120         79    2625          18.6  F-1694136


<font size = "4">

- Grab mpg, weight, vehicle_id for **all** cars (columns 0, 4, 6)

In [18]:
subfeatures = carfeatures.iloc[ : , [0, 4, 6] ]
print(subfeatures)

      mpg  weight vehicle_id
0    18.0    3504  C-1689780
1    15.0    3693  B-1689791
2    18.0    3436  P-1689802
3    16.0    3433  A-1689813
4    17.0    3449  F-1689824
..    ...     ...        ...
393  27.0    2790  F-1694103
394  44.0    2130  V-1694114
395  32.0    2295  D-1694125
396  28.0    2625  F-1694136
397  31.0    2720  C-1694147

[398 rows x 3 columns]


In [19]:
# Grab mpg, weight, vehicle_id for ALL cars (columns 0, 4, 6)
sub_cols = [0, 4, 6]
subfeatures = carfeatures.iloc[ : , sub_cols]
print(subfeatures)

      mpg  weight vehicle_id
0    18.0    3504  C-1689780
1    15.0    3693  B-1689791
2    18.0    3436  P-1689802
3    16.0    3433  A-1689813
4    17.0    3449  F-1689824
..    ...     ...        ...
393  27.0    2790  F-1694103
394  44.0    2130  V-1694114
395  32.0    2295  D-1694125
396  28.0    2625  F-1694136
397  31.0    2720  C-1694147

[398 rows x 3 columns]


<font size = "4">

- Grab mpg, weight, vehicle_id for cars 0, 4, 396

- We need rows 0, 4, 396

- And we need columns 0, 4, 6

In [20]:
subdata = carfeatures.iloc[ [0, 4, 396], [0, 4, 6] ]
print(subdata)

      mpg  weight vehicle_id
0    18.0    3504  C-1689780
4    17.0    3449  F-1689824
396  28.0    2625  F-1694136


### Subsetting blocks of rows

<font size = "4">

- Let's grab all features for the first 5 cars (rows 0, 1, 2, 3, 4)

In [21]:
first_5_cars = carfeatures.iloc[ :5, :]
print(first_5_cars)

    mpg  cylinders  displacement horsepower  weight  acceleration vehicle_id
0  18.0          8           307        130    3504          12.0  C-1689780
1  15.0          8           350        165    3693          11.5  B-1689791
2  18.0          8           318        150    3436          11.0  P-1689802
3  16.0          8           304        150    3433          12.0  A-1689813
4  17.0          8           302        140    3449          10.5  F-1689824


<font size = "4">

- The next cell grabs the features for the first 5 cars in an alternate way.

- We start with row 0

- The first row we **exclude** is row 5

In [28]:
first_5_cars = carfeatures.iloc[0:5, :]
print(first_5_cars)

    mpg  cylinders  displacement horsepower  weight  acceleration vehicle_id
0  18.0          8           307        130    3504          12.0  C-1689780
1  15.0          8           350        165    3693          11.5  B-1689791
2  18.0          8           318        150    3436          11.0  P-1689802
3  16.0          8           304        150    3433          12.0  A-1689813
4  17.0          8           302        140    3449          10.5  F-1689824


<font size = "4">

- What happens if we shift the indices by 1 (change 0:5 --> 1:6)?

- Then we will grab rows (1, 2, 3, 4, 5)

- So we start with row 1 and the first row we **exclude** is row 6.

In [24]:
cars_1_thru_5 = carfeatures.iloc[ 1:6, :]

print(cars_1_thru_5)

    mpg  cylinders  displacement horsepower  weight  acceleration vehicle_id
1  15.0          8           350        165    3693          11.5  B-1689791
2  18.0          8           318        150    3436          11.0  P-1689802
3  16.0          8           304        150    3433          12.0  A-1689813
4  17.0          8           302        140    3449          10.5  F-1689824
5  15.0          8           429        198    4341          10.0  F-1689835


<font size = "4">

- Grab rows 211-219 (so 220 is the first one we exclude)

In [29]:

print(carfeatures.iloc[ 211:220 , :])

      mpg  cylinders  displacement horsepower  weight  acceleration vehicle_id
211  16.5          6           168        120    3820          16.7  M-1692101
212  16.5          8           350        180    4380          12.1  C-1692112
213  13.0          8           350        145    4055          12.0  C-1692123
214  13.0          8           302        130    3870          15.0  F-1692134
215  13.0          8           318        150    3755          14.0  D-1692145
216  31.5          4            98         68    2045          18.5  H-1692156
217  30.0          4           111         80    2155          14.8  B-1692167
218  36.0          4            79         58    1825          18.6  R-1692178
219  25.5          4           122         96    2300          15.5  P-1692189


In [32]:
print(carfeatures.iloc[87:101,])

      mpg  cylinders  displacement horsepower  weight  acceleration vehicle_id
87   13.0          8           350        145    3988          13.0  C-1690737
88   14.0          8           302        137    4042          14.5  F-1690748
89   15.0          8           318        150    3777          12.5  D-1690759
90   12.0          8           429        198    4952          11.5  M-1690770
91   13.0          8           400        150    4464          12.0  C-1690781
92   13.0          8           351        158    4363          13.0  F-1690792
93   14.0          8           318        150    4237          14.5  P-1690803
94   13.0          8           440        215    4735          11.0  C-1690814
95   12.0          8           455        225    4951          11.0  B-1690825
96   13.0          8           360        175    3821          11.0  A-1690836
97   18.0          6           225        105    3121          16.5  P-1690847
98   16.0          6           250        100    327

<font size = "4">

- Let's grab the last 3 rows (rows 395, 396, 397)

In [33]:
# grab last 3 rows (rows 395, 396, and 397)
print(carfeatures.iloc[ -3: , :])

      mpg  cylinders  displacement horsepower  weight  acceleration vehicle_id
395  32.0          4           135         84    2295          11.6  D-1694125
396  28.0          4           120         79    2625          18.6  F-1694136
397  31.0          4           119         82    2720          19.4  C-1694147


<font size = "4">

- Let's grab the 3rd from last row and the 2nd from last row.

- So we start at row "-3" and the first row we exclude is row "-1"

In [34]:
print(carfeatures.iloc[ -3:-1 , :])

      mpg  cylinders  displacement horsepower  weight  acceleration vehicle_id
395  32.0          4           135         84    2295          11.6  D-1694125
396  28.0          4           120         79    2625          18.6  F-1694136


<font size = "4">

Works similarly for blocks of columns, or blocks of both rows and columns

**Exercise:** Extract the columns corresponding to displacement, horsepower, and weight for all cars using ``.iloc``. The column indices should be of the form ``start:stop`` for appropriate integers ``start`` and ``stop``

**Exercise:** Extract the weight, acceleration, and vehicle_id information for cars 28, 29, 30, 31, 32 using ":" for both row and column indices

In [ ]:
## your answer here



<font size = "4">

This exercise demonstrates an important application: using `.iloc` in tandem with the `.sort_values()` method.

**Exercise:** Create a dataset ``car_ascendingweight`` which sorts the cars from lowest to highest weight. Then extract the subset of the data corresponding to the 5 cars with the lowest weight (all their features, not just their weight)

In [39]:
# your answer here
car_ascendingweight = carfeatures.sort_values("weight")
print(car_ascendingweight)

print(car_ascendingweight.iloc[0:5,])

      mpg  cylinders  displacement horsepower  weight  acceleration vehicle_id
54   35.0          4            72         69    1613          18.0  D-1690374
144  31.0          4            76         52    1649          16.5  T-1691364
343  39.1          4            79         58    1755          16.9  T-1693553
345  35.1          4            81         60    1760          16.1  H-1693575
53   31.0          4            71         65    1773          19.0  T-1690363
..    ...        ...           ...        ...     ...           ...        ...
95   12.0          8           455        225    4951          11.0  B-1690825
90   12.0          8           429        198    4952          11.5  M-1690770
42   12.0          8           383        180    4955          11.5  D-1690242
103  11.0          8           400        150    4997          14.0  C-1690913
44   13.0          8           400        175    5140          12.0  P-1690264

[398 rows x 7 columns]
      mpg  cylinders  displa

## <span style="color:red"> III. Adding Variables </span>

<font size = "4">

- Sometimes we don't want to simply read in a dataset, but **manipulate** the data in some way
- One of the most simple data manipulations is to add new variables/columns

We'll demonstrate with a data-set of the heights and weights of a sample of people. The dataset is from the Statistics Online Computational Resource (SOCR)

In [40]:
df_ht_wt = pd.read_csv("data/SOCR-HeightWeight.csv")
print(df_ht_wt.columns.values)
print()
print(df_ht_wt)

['Height(Inches)' 'Weight(Pounds)']

       Height(Inches)  Weight(Pounds)
0            65.78331        112.9925
1            71.51521        136.4873
2            69.39874        153.0269
3            68.21660        142.3354
4            67.78781        144.2971
...               ...             ...
24995        69.50215        118.0312
24996        64.54826        120.1932
24997        64.69855        118.2655
24998        67.52918        132.2682
24999        68.87761        124.8742

[25000 rows x 2 columns]


<font size = "4">

- The dataset contains the height and weight of people measured in inches and pounds, respectively. 
- Using these two variables, we can calculate a new variable: the BMI (Body Mass Index) of each person.
- The formula for BMI is given by the formula:
$$\textrm{BMI } = \frac{\textrm{weight in kg}}{\left(\textrm{height in meters}\right)^2}$$

In [45]:
# Grab height and weight and convert to correct units

def pounds_to_kg(wt):
    return wt*0.45359237

def inches_to_meters(ht):
    ht_mm = ht*25.4
    return ht_mm*1e-3

ht_inch = df_ht_wt["Height(Inches)"]
wt_lbs = df_ht_wt["Weight(Pounds)"]

ht_meters = ht_inch.apply(inches_to_meters)
ht_meters.name = "Height(Meters)"
wt_kg = wt_lbs.apply(pounds_to_kg)
wt_kg.name = "Weight(Kilograms)"

print(wt_kg)
print(ht_meters)

0        51.252536
1        61.909598
2        69.411834
3        64.562251
4        65.452064
           ...    
24995    53.538052
24996    54.518718
24997    53.644328
24998    59.995846
24999    56.641984
Name: Weight(Kilograms), Length: 25000, dtype: float64
0        1.670896
1        1.816486
2        1.762728
3        1.732702
4        1.721810
           ...   
24995    1.765355
24996    1.639526
24997    1.643343
24998    1.715241
24999    1.749491
Name: Height(Meters), Length: 25000, dtype: float64


<font size = "4">

**Exercise**: Repeat the operations in the last cell, but grab ``ht_inch`` and ``wt_lbs`` using ``.iloc`` instead of the column names

In [46]:
# your code here



<font size = "4">

Now, calculate BMI using these two variables and add to the dataset

In [47]:
df_ht_wt.head()

,Height(Inches),Weight(Pounds)
0,65.78331,112.9925
1,71.51521,136.4873
2,69.39874,153.0269
3,68.21660,142.3354
4,67.78781,144.2971


In [50]:
# Math operations on Pandas series work very similarly to Numpy arrays

BMI = wt_kg/ht_meters**2
print(BMI)
# assign the new Pandas series to a new column of the dataframe
df_ht_wt["BMI"] = BMI

0        18.357624
1        18.762630
2        22.338958
3        21.504586
4        22.077643
           ...    
24995    17.179031
24996    20.281923
24997    19.864026
24998    20.392515
24999    18.506099
Length: 25000, dtype: float64


In [49]:
df_ht_wt.head()

,Height(Inches),Weight(Pounds),BMI
0,65.78331,112.9925,18.357624
1,71.51521,136.4873,18.762630
2,69.39874,153.0269,22.338958
3,68.21660,142.3354,21.504586
4,67.78781,144.2971,22.077643


<font size = "4">

**Application:** Random Assignment

- Random assignment is an experimental technique for dividing a data set into groups using randomization
- Typically a data point has an equal probability of being placed in each group (but not necessary)
- Let's say we want to divide up the height/weight dataset into 4 groups. We can use Numpy.random to assign each person the number 1, 2, 3, or 4

In [ ]:
import numpy as np

# get size of data-set
num_rows, num_cols = df_ht_wt.shape

# randomly generate a number (1,2,3, or 4) for each person in the dataset
group_no = np.random.choice([1, 2, 3, 4], size = num_rows)

df_ht_wt["GroupID"] = group_no


In [ ]:
df_ht_wt.head()

## <span style="color:red"> IV. Filtering Dataframes Based on Logical Expressions </span>

<font size = "4">

(i) Filtering using ```pandas.query()``` 

```python
data.query("logical expression")
```

In [ ]:
high_mileagecars = carfeatures.query("mpg >= 25")
print(high_mileagecars)

In [ ]:
group_4 = df_ht_wt.query("GroupID == 4")
print(group_4)

In [ ]:
# Use "and" / "or" for logical operations on queries

car_subset = carfeatures.query("(acceleration >= 12) and (acceleration < 18)")
print(car_subset)

<font size = "4">

(ii) Expressions with colnames + global variables (@)

In [ ]:
# You can invoke global variables into the query by using @variablename
# If you don't include @, then Python will try to look for a column with 
# that name.

threshold = 25
high_mileagecars = carfeatures.query("mpg >= @threshold")
print(high_mileagecars)

In [ ]:
lb = 12
ub = 18

car_subset = carfeatures.query("(acceleration >= @lb) and (acceleration < @ub)")
print(car_subset)

<font size = "4">

(iii) Handling white space

Ever wonder why I give folders names like "lecture_18" instead of "lecture 18"? Or name notebooks things like "lecture_18a_pandas_operations.ipynb" instead of "lecture 18a - Pandas operations.ipynb"?

Because handling white spaces on a computer can be annoying in certain situations, as we will see here:

In [ ]:
# same data set, but column/variable 0 is "miles per gallon" instead of "mpg"

new_dset = pd.read_csv("data/features-bad-spacing.csv")
print(new_dset)

In [ ]:
data_threshold_horsepower = new_dset.query("miles per gallon >= 20")

<font size = "4">

Pretty scary looking error message...basically, it doesn't like the space (variable names can't have spaces in Python)

We've seen the double quote (") and the single quote (')

To the left of your 1 key is the backtick/grave accent (`). Need to use this

In [ ]:
# next line causes error
# data_threshold_horsepower = new_dset.query("miles per gallon >= 20")

# this next line is fine
data_threshold_horsepower = new_dset.query("`miles per gallon` >= 20")
print(data_threshold_horsepower)

<font size = "4">

**Exercise:** Subset the data, gathering all cars with exactly 8 cylinders and mpg $\ge$ 25.


In [ ]:
# Your solution here




## <span style="color:red"> VII. Application: Visualizing Subsets of the Data </span>

<font size = "4">

List of unique categories

In [ ]:
# Use pd.unique() to extract a numpy array with the unique elements in that column

array_unique_cylinders = pd.unique(carfeatures["cylinders"])
print(array_unique_cylinders)
print(type(array_unique_cylinders))

<font size = "4">

Compute two overlapping plots

In [ ]:
# If we call plt.scatter() twice to display two plots
# To display all plots simultaneously we include plt.show() at the very end.
# The idea is that the graphs are stacked on top of each other

df_8 = carfeatures.query("cylinders == 8")
df_4 = carfeatures.query("cylinders == 4")

plt.scatter(x = df_8["weight"],y = df_8["acceleration"])
plt.scatter(x = df_4["weight"],y = df_4["acceleration"])
plt.legend(labels = ["8","4"],
           title  = "Cylinders")

plt.xlabel("Weight")
plt.ylabel("Acceleration")
plt.show()

# Note: If we put plt.show() in between the plots, then the results will
# be shown on separate graphs instead.

<font size = "4">

Compute plots by all categories

In [ ]:
# Compute number of unique categories
list_unique_cylinders = pd.unique(carfeatures["cylinders"])

# Use a for loop to plot a scatter plot between "weight" and "acceleration"
# for each category. Each plot  will have a different color

for category in list_unique_cylinders:
    df   = carfeatures.query("cylinders == @category")
    plt.scatter(x = df["weight"], y = df["acceleration"])
    
# Add labels and a legends    
plt.xlabel("Weight")
plt.ylabel("Acceleration")
plt.legend(labels = list_unique_cylinders,
           title  = "Cylinders")
plt.show()

<font size="4">

**Exercise:** Compute a histogram of "mpg" by cylinder count for cars with 4, 6, and 8 cylinders.Here is what it might look like (using a fake dataset):

<img src="data/fake_histogram.png" alt="drawing" width="400"/>


You can make the histograms transparent by adjusting ```alpha``` in

```python
    plt.hist(x = ..., alpha = 0.5)
```

In [ ]:
# your code here

